examples/medplum-demo-bots/src/deduplication/merge-matching-patients.ts

```diff
@@ -0,0 +1,199 @@
+  }
+
+  // Update the target patient record with the merged data, and have it as the master record.
+  await medplum.updateResource<Patient>(mergedPatients.target);
+}
+
+interface MergedPatients {
+  readonly src: Patient;
+  readonly target: Patient;
+}
+
+type HasSubject = {
+  subject?: Reference<Patient>;
+};
+
+type ResourceWithSubject = Resource & HasSubject;
+
+/**
+ * Links two patient records indicating one replaces the other.
+ *
+ * @param src - The source patient record which is being replaced.
+ * @param target - The target patient record which will replace the source.
+ * @returns - Object containing updated source and target patient records with their links.
+ */
+export function linkPatientRecords(src: Patient, target: Patient): MergedPatients {
+  const targetCopy = deepClone(target);
+  const targetLinks = targetCopy.link ?? [];
+  targetLinks.push({ other: createReference(src), type: 'replaces' });
+
+  const srcCopy = deepClone(src);
+  const srcLinks = srcCopy.link ?? [];
+  srcLinks.push({ other: createReference(target), type: 'replaced-by' });
+  return { src: { ...srcCopy, link: srcLinks, active: false }, target: { ...targetCopy, link: targetLinks } };
+}
+
+/**
+ * Merges contact information (identifiers) of two patient records, where the source patient record will be marked as an old record.
+ * The target patient record will be overwritten with the merged data and will be the master record.
+ *
+ * @param src - The source patient record.
+ * @param target - The target patient record.
+ * @param fields - Optional additional fields to be merged.
+ * @returns - Object containing the original source and the merged target patient records.
+ */
+export function mergePatientRecords(src: Patient, target: Patient, fields?: Partial<Patient>): MergedPatients {
+  const srcIdentifiers = src.identifier ?? [];
+  const mergedIdentifiers = srcIdentifiers.map((identifier) => ({
+    ...identifier,
+    use: 'old' as Identifier['use'],
+  }));
+  const targetCopy = deepClone(target);
+  targetCopy.identifier = [...(targetCopy.identifier ?? []), ...mergedIdentifiers];
+  const targetMerged = { ...targetCopy, ...fields };
+  return { src: src, target: targetMerged };
+}
+
+/**
+ * Updates the subject of clinical resources from the source patient to the target patient.
+ *
+ * @param medplum - Instance of the MedplumClient for server interactions.
+ * @param src - The source patient record.
+ * @param target - The target patient record.
+ * @param clinicalResource - The type of the clinical resource (e.g., 'ServiceRequest').
+ */
+export async function updateClinicalReferences<T extends ResourceWithSubject>(
+  medplum: MedplumClient,
+  src: Patient,
+  target: Patient,
+  clinicalResource: T['resourceType']
+): Promise<void> {
+  // Search for clinical resources related to the source patient.
+  const reports = await medplum.searchResources(clinicalResource, { subject: getReferenceString(src) });
```

Comment:
Just searching on `subject` might not be sufficient; we should use the Patient compartment to ensure all relevant resources are found via search

Reply:
Good catch!